### DISCUSSION TITLE
Feature to track files where citation is used

### DISCUSSION THREAD

**Discussion by nicolas-graves** [emacs-citar/citar#816] [Ideas]
Thanks for this package. Has anyone already worked / documented such a feature:
Let's say you have ~20 files you produce a year with citations. You want to be able to reproduce and update them easily. But you also want to be able to edit your one and only bibliographic file if you either change the style of keys or update the name of a key. Having a cache / functionality that would be able to append for any citation where / in which document it is cited would greatly help. 
This could happen on write (simply a hook when you save such a file).

Is anyone also interested in such a feature ?

**Comment by bdarcus** [comment]
To clarify @nicolas-graves, two use cases/requirements?

1. Identify files which include specific citations ("You want to be able to reproduce and update them easily.")
2. Somehow sync the citation keys in documents and in the bib file(s) if they change ("But you also want to be able to edit your one and only bibliographic file if you either change the style of keys or update the name of a key.")

Yes, they overlap, but they're also distinct?

If that's right, I'd think the first one would be more straightforward and general. It would basically be an extension to what we do now with related resources. So you could imagine doing `M-x citar-open` and there being a section for "Cited In". See also for similar idea for notes: https://github.com/emacs-citar/citar-org-roam/issues/40.

The second one, however, sounds pretty user-specific? Or at least, it's not clear to me how you see that second one working specifically. Care to clarify?

**Reply by nicolas-graves** [reply]
Thanks for the quick answer @bdarcus 
Yes, I can clarify. 
I agree that there are two functionalities, and that they overlap. The part about the change of the bibliographic file can be put aside / is indeed a special case of persistence.

To me, the biggest question to clarify the (2)nd point is: What kind of cache / persistence we want to use?
a) no persistence: This amounts to the (1)st point and an approach that would probably make a directory variable, and parse each org file to find citations, record them in a hash-table / alist with pairs `(citar-key . ,(list "filea" "fileb"))`. This can get computationally intensive.
b) emacs-specific persistence: This is (a) + use of an emacs cache that allows persistence. 
  i) My approach would be to have a on-save file hook to append the file the aformentioned alist as I do in https://github.com/nicolas-graves/org-agenda-files-track (I'm biased of course). This doesn't require having a "hard" directory requirement on the user such as a directory variable (directory to put citing files in). 
  ii) Some modern caching approaches as org-ql and possibly citar are doing might not really need this kind of hook since they are aware what files / part of files have changed, but this is hard  to imagine without a directory variable.
c) general persistence: This implies data is additionally saved independently of emacs, and described as your (2). 

The (c) is of course much more "user-specific" if we want, though I see portability and bibliographic file unity as a wanted feature. But anyway, you're right, we can for instance consider (b)(i) or (b)(ii) should be a citar feature and (c)  could be a wiki script made by an user to pass the data persistence in an emacs-independent form.
